File: backend/src/main/java/com/anushibinj/mypromptlib/service/UserService.java
package com.anushibinj.mypromptlib.service;

import com.anushibinj.mypromptlib.model.User;
import com.anushibinj.mypromptlib.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;

    public User registerUser(String username, String password) {
        if (userRepository.findByUsername(username).isPresent()) {
            throw new RuntimeException("Username already exists");
        }
        User user = User.builder()
                .username(username)
                .password(password)
                .token(UUID.randomUUID().toString())
                .build();
        return userRepository.save(user);
    }

    public Optional<User> loginUser(String username, String password) {
        Optional<User> userOptional = userRepository.findByUsername(username);
        if (userOptional.isPresent() && password != null && password.equals(userOptional.get().getPassword())) {
            User user = userOptional.get();
            // Assign new token
            user.setToken(UUID.randomUUID().toString());
            userRepository.save(user);
            return Optional.of(user);
        }
        return Optional.empty();
    }

    public User loginOrRegisterGoogleUser(String googleId, String email) {
        // Check if user already linked to this Google account
        Optional<User> existingGoogleUser = userRepository.findByGoogleId(googleId);
        if (existingGoogleUser.isPresent()) {
            User user = existingGoogleUser.get();
            user.setToken(UUID.randomUUID().toString());
            return userRepository.save(user);
        }
        // Check if user exists with same email as username — link Google ID
        Optional<User> emailUser = userRepository.findByUsername(email);
        if (emailUser.isPresent()) {
            User user = emailUser.get();
            user.setGoogleId(googleId);
            user.setToken(UUID.randomUUID().toString());
            return userRepository.save(user);
        }
        // Create new Google user
        User newUser = User.builder()
                .username(email)
                .googleId(googleId)
                .token(UUID.randomUUID().toString())
                .build();
        return userRepository.save(newUser);
    }

    public Optional<User> findByToken(String token) {
        if (token == null || token.isBlank()) return Optional.empty();
        return userRepository.findByToken(token);
    }
}

File: backend-go/internal/service/user_service.go
package service

import (
	"errors"
	"backend-go/internal/model"
	"backend-go/internal/repository"
	"github.com/google/uuid"
)

type UserService struct {
	repo *repository.UserRepository
}

func NewUserService(repo *repository.UserRepository) *UserService {
	return &UserService{repo: repo}
}

func (s *UserService) RegisterUser(username, password string) (*model.User, error) {
	existingUser, err := s.repo.FindByUsername(username)
	if err != nil {
		return nil, err
	}
	if existingUser != nil {
		return nil, errors.New("Username already exists")
	}

	user := &model.User{
		ID:       uuid.New(),
		Username: username,
		Password: password,
		Token:    uuid.New().String(),
	}

	return s.repo.Save(user)
}

func (s *UserService) LoginUser(username, password string) (*model.User, error) {
	user, err := s.repo.FindByUsername(username)
	if err != nil {
		return nil, err
	}
	if user != nil && password != "" && user.Password == password {
		// Assign new token
		user.Token = uuid.New().String()
		return s.repo.Save(user)
	}
	return nil, nil // Return nil, nil when login fails (like Optional.empty())
}

func (s *UserService) LoginOrRegisterGoogleUser(googleID, email string) (*model.User, error) {
	// Check if user already linked to this Google account
	existingGoogleUser, err := s.repo.FindByGoogleID(googleID)
	if err != nil {
		return nil, err
	}
	if existingGoogleUser != nil {
		existingGoogleUser.Token = uuid.New().String()
		return s.repo.Save(existingGoogleUser)
	}

	// Check if user exists with same email as username — link Google ID
	emailUser, err := s.repo.FindByUsername(email)
	if err != nil {
		return nil, err
	}
	if emailUser != nil {
		emailUser.GoogleID = googleID
		emailUser.Token = uuid.New().String()
		return s.repo.Save(emailUser)
	}

	// Create new Google user
	newUser := &model.User{
		ID:       uuid.New(),
		Username: email,
		GoogleID: googleID,
		Token:    uuid.New().String(),
	}
	return s.repo.Save(newUser)
}

func (s *UserService) FindByToken(token string) (*model.User, error) {
	if token == "" {
		return nil, nil
	}
	return s.repo.FindByToken(token)
}
